File: python_source/expense_tracker.py
from datetime import datetime
from pathlib import Path
import csv


DATA_FILE = Path("data/expenses.csv")
DATE_FORMAT = "%Y-%m-%d"


def create_expense(amount, category, description, date):
    return {
        "amount": amount,
        "category": category,
        "description": description,
        "date": date,
    }


def parse_date(date_text):
    try:
        parsed_date = datetime.strptime(date_text, DATE_FORMAT).date()
    except ValueError:
        return None

    if parsed_date.strftime(DATE_FORMAT) != date_text:
        return None

    return parsed_date


def prompt_for_amount(prompt, allow_blank=False):
    while True:
        value = input(prompt).strip()

        if allow_blank and value == "":
            return None

        try:
            amount = float(value)
        except ValueError:
            print("Invalid amount. Please enter a number.")
            continue

        if amount < 0:
            print("Amount cannot be negative.")
            continue

        return amount


def prompt_for_date(prompt, allow_blank=False):
    while True:
        value = input(prompt).strip()

        if allow_blank and value == "":
            return ""

        if parse_date(value) is None:
            print("Invalid date. Please use YYYY-MM-DD.")
            continue

        return value


def prompt_for_required_text(prompt):
    while True:
        value = input(prompt).strip()

        if value:
            return value

        print("This field cannot be blank.")


def prompt_for_expense_number(expenses):
    if not expenses:
        print("No expenses recorded.")
        return None

    view_expenses(expenses)

    try:
        number = int(input("Expense number: ").strip())
    except ValueError:
        print("Invalid expense number.")
        return None

    if number < 1 or number > len(expenses):
        print("Expense number out of range.")
        return None

    return number - 1


def add_expense(expenses):
    print("\nAdd Expense")

    amount = prompt_for_amount("Amount: $")
    category = prompt_for_required_text("Category: ")
    description = prompt_for_required_text("Description: ")
    date = prompt_for_date("Date (YYYY-MM-DD): ")

    expense = create_expense(amount, category, description, date)
    expenses.append(expense)
    print("Expense added.")


def edit_expense(expenses):
    print("\nEdit Expense")
    index = prompt_for_expense_number(expenses)

    if index is None:
        return

    expense = expenses[index]
    print("Press Enter to keep the current value.")

    amount = prompt_for_amount(
        f"Amount (${expense['amount']:.2f}): ",
        allow_blank=True,
    )
    category = input(f"Category ({expense['category']}): ").strip()
    description = input(f"Description ({expense['description']}): ").strip()
    date = prompt_for_date(
        f"Date ({expense['date']}): ",
        allow_blank=True,
    )

    if amount is not None:
        expense["amount"] = amount
    if category:
        expense["category"] = category
    if description:
        expense["description"] = description
    if date:
        expense["date"] = date

    print("Expense updated.")


def delete_expense(expenses):
    print("\nDelete Expense")
    index = prompt_for_expense_number(expenses)

    if index is None:
        return

    expense = expenses[index]
    print(
        f"Delete ${expense['amount']:.2f} | "
        f"{expense['category']} | "
        f"{expense['description']} | "
        f"{expense['date']}?"
    )
    confirm = input("Type yes to delete: ").strip().lower()

    if confirm == "yes":
        expenses.pop(index)
        print("Expense deleted.")
    else:
        print("Delete canceled.")


def format_expense(index, expense):
    return (
        f"{index}. ${expense['amount']:.2f} | "
        f"{expense['category']} | "
        f"{expense['description']} | "
        f"{expense['date']}"
    )


def view_expenses(expenses, title="All Expenses"):
    print(f"\n{title}")

    if not expenses:
        print("No expenses recorded.")
        return

    for index, expense in enumerate(expenses, start=1):
        print(format_expense(index, expense))


def search_by_category(expenses):
    print("\nSearch By Category")
    category = input("Enter category: ").strip().lower()

    matches = [
        expense
        for expense in expenses
        if expense["category"].lower() == category
    ]

    if not matches:
        print("No expenses found in that category.")
        return

    view_expenses(matches, "Matching Expenses")


def filter_by_date_range(expenses):
    print("\nFilter By Date Range")
    start_text = prompt_for_date("Start date (YYYY-MM-DD): ")
    end_text = prompt_for_date("End date (YYYY-MM-DD): ")
    start_date = parse_date(start_text)
    end_date = parse_date(end_text)

    if start_date > end_date:
        print("Start date cannot be after end date.")
        return

    matches = []

    for expense in expenses:
        expense_date = parse_date(expense["date"])

        if expense_date is None:
            continue

        if start_date <= expense_date <= end_date:
            matches.append(expense)

    if not matches:
        print("No expenses found in that date range.")
        return

    view_expenses(matches, f"Expenses From {start_text} To {end_text}")


def sort_expenses(expenses):
    print("\nSort Expenses")

    if not expenses:
        print("No expenses recorded.")
        return

    print("1. Date")
    print("2. Amount")
    print("3. Category")
    print("4. Description")
    choice = input("Sort by: ").strip()
    reverse = input("Descending? (y/n): ").strip().lower() == "y"

    if choice == "1":
        sorted_expenses = sorted(
            expenses,
            key=lambda expense: parse_date(expense["date"]),
            reverse=reverse,
        )
        title = "Expenses Sorted By Date"
    elif choice == "2":
        sorted_expenses = sorted(
            expenses,
            key=lambda expense: expense["amount"],
            reverse=reverse,
        )
        title = "Expenses Sorted By Amount"
    elif choice == "3":
        sorted_expenses = sorted(
            expenses,
            key=lambda expense: expense["category"].lower(),
            reverse=reverse,
        )
        title = "Expenses Sorted By Category"
    elif choice == "4":
        sorted_expenses = sorted(
            expenses,
            key=lambda expense: expense["description"].lower(),
            reverse=reverse,
        )
        title = "Expenses Sorted By Description"
    else:
        print("Invalid sort option.")
        return

    view_expenses(sorted_expenses, title)


def show_total(expenses):
    total = sum(expense["amount"] for expense in expenses)
    print(f"\nTotal spent: ${total:.2f}")


def show_totals_by_category(expenses):
    print("\nTotals By Category")

    if not expenses:
        print("No expenses recorded.")
        return

    totals = {}

    for expense in expenses:
        category = expense["category"]
        totals[category] = totals.get(category, 0) + expense["amount"]

    for category in sorted(totals):
        print(f"{category}: ${totals[category]:.2f}")


def save_expenses(expenses):
    DATA_FILE.parent.mkdir(exist_ok=True)

    with DATA_FILE.open("w", newline="") as file:
        writer = csv.DictWriter(
            file,
            fieldnames=["amount", "category", "description", "date"],
        )
        writer.writeheader()
        writer.writerows(expenses)

    print(f"Saved {len(expenses)} expenses to {DATA_FILE}.")


def load_expenses():
    expenses = []

    if not DATA_FILE.exists():
        return expenses

    with DATA_FILE.open("r", newline="") as file:
        reader = csv.DictReader(file)

        for row in reader:
            try:
                amount = float(row.get("amount", ""))
            except ValueError:
                continue

            category = row.get("category", "").strip()
            description = row.get("description", "").strip()
            date = row.get("date", "").strip()

            if amount < 0 or not category or not description:
                continue

            if parse_date(date) is None:
                continue

            expenses.append(create_expense(amount, category, description, date))

    return expenses


def print_menu():
    print("\nExpense Tracker")
    print("1. Add expense")
    print("2. Edit expense")
    print("3. Delete expense")
    print("4. View all expenses")
    print("5. Search by category")
    print("6. Filter by date range")
    print("7. Sort expenses")
    print("8. Show total spent")
    print("9. Show totals by category")
    print("10. Save expenses")
    print("11. Load expenses")
    print("12. Exit")


def main():
    expenses = load_expenses()
    print(f"Loaded {len(expenses)} expenses.")

    while True:
        print_menu()
        choice = input("Choose an option: ").strip()

        if choice == "1":
            add_expense(expenses)
        elif choice == "2":
            edit_expense(expenses)
        elif choice == "3":
            delete_expense(expenses)
        elif choice == "4":
            view_expenses(expenses)
        elif choice == "5":
            search_by_category(expenses)
        elif choice == "6":
            filter_by_date_range(expenses)
        elif choice == "7":
            sort_expenses(expenses)
        elif choice == "8":
            show_total(expenses)
        elif choice == "9":
            show_totals_by_category(expenses)
        elif choice == "10":
            save_expenses(expenses)
        elif choice == "11":
            expenses = load_expenses()
            print(f"Loaded {len(expenses)} expenses.")
        elif choice == "12":
            save_expenses(expenses)
            print("Goodbye.")
            break
        else:
            print("Invalid choice. Please choose 1-12.")


if __name__ == "__main__":
    main()

File: java_translation/src/ExpenseTracker.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.time.LocalDate;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Scanner;
import java.util.TreeMap;

public class ExpenseTracker {
    private static final String DATA_FILE = "data/expenses.csv";

    private ArrayList<Expense> expenses;
    private Scanner scanner;

    public ExpenseTracker() {
        expenses = new ArrayList<>();
        scanner = new Scanner(System.in);
    }

    public void run() {
        loadExpenses();
        System.out.println("Loaded " + expenses.size() + " expenses.");

        while (true) {
            printMenu();
            System.out.print("Choose an option: ");
            String choice = scanner.nextLine().trim();

            if (choice.equals("1")) {
                addExpense();
            } else if (choice.equals("2")) {
                editExpense();
            } else if (choice.equals("3")) {
                deleteExpense();
            } else if (choice.equals("4")) {
                viewExpenses(expenses);
            } else if (choice.equals("5")) {
                searchByCategory();
            } else if (choice.equals("6")) {
                filterByDateRange();
            } else if (choice.equals("7")) {
                sortExpenses();
            } else if (choice.equals("8")) {
                showTotal();
            } else if (choice.equals("9")) {
                showTotalsByCategory();
            } else if (choice.equals("10")) {
                saveExpenses();
            } else if (choice.equals("11")) {
                loadExpenses();
                System.out.println("Loaded " + expenses.size() + " expenses.");
            } else if (choice.equals("12")) {
                saveExpenses();
                System.out.println("Goodbye. Thanks for using this expense tracker by the handsome Khanh!");
                break;
            } else {
                System.out.println("Invalid choice. Please choose 1-12.");
            }
        }
    }

    private void printMenu() {
        System.out.println("\nExpense Tracker");
        System.out.println("1. Add expense");
        System.out.println("2. Edit expense");
        System.out.println("3. Delete expense");
        System.out.println("4. View all expenses");
        System.out.println("5. Search by category");
        System.out.println("6. Filter by date range");
        System.out.println("7. Sort expenses");
        System.out.println("8. Show total spent");
        System.out.println("9. Show totals by category");
        System.out.println("10. Save expenses");
        System.out.println("11. Load expenses");
        System.out.println("12. Exit");
    }

    private LocalDate parseDate(String dateText) {
        try {
            return LocalDate.parse(dateText);
        } catch (DateTimeParseException error) {
            return null;
        }
    }

    private Double promptForAmount(String prompt, boolean allowBlank) {
        while (true) {
            System.out.print(prompt);
            String value = scanner.nextLine().trim();

            if (allowBlank && value.equals("")) {
                return null;
            }

            try {
                double amount = Double.parseDouble(value);

                if (amount < 0) {
                    System.out.println("Amount cannot be negative.");
                    continue;
                }

                return amount;
            } catch (NumberFormatException error) {
                System.out.println("Invalid amount. Please enter a number.");
            }
        }
    }

    private String promptForDate(String prompt, boolean allowBlank) {
        while (true) {
            System.out.print(prompt);
            String value = scanner.nextLine().trim();

            if (allowBlank && value.equals("")) {
                return "";
            }

            if (parseDate(value) == null) {
                System.out.println("Invalid date. Please use YYYY-MM-DD.");
                continue;
            }

            return value;
        }
    }

    private String promptForRequiredText(String prompt) {
        while (true) {
            System.out.print(prompt);
            String value = scanner.nextLine().trim();

            if (!value.equals("")) {
                return value;
            }

            System.out.println("This field cannot be blank.");
        }
    }

    private int promptForExpenseIndex() {
        if (expenses.isEmpty()) {
            System.out.println("No expenses recorded.");
            return -1;
        }

        viewExpenses(expenses);
        System.out.print("Expense number: ");

        int number;
        try {
            number = Integer.parseInt(scanner.nextLine().trim());
        } catch (NumberFormatException error) {
            System.out.println("Invalid expense number.");
            return -1;
        }

        if (number < 1 || number > expenses.size()) {
            System.out.println("Expense number out of range.");
            return -1;
        }

        return number - 1;
    }

    private void addExpense() {
        System.out.println("\nAdd Expense");

        double amount = promptForAmount("Amount: $", false);
        String category = promptForRequiredText("Category: ");
        String description = promptForRequiredText("Description: ");
        String date = promptForDate("Date (YYYY-MM-DD): ", false);

        Expense expense = new Expense(amount, category, description, date);
        expenses.add(expense);
        System.out.println("Expense added.");
    }

    private void editExpense() {
        System.out.println("\nEdit Expense");
        int index = promptForExpenseIndex();

        if (index == -1) {
            return;
        }

        Expense expense = expenses.get(index);
        System.out.println("Press Enter to keep the current value.");

        Double amount = promptForAmount(
            String.format("Amount ($%.2f): ", expense.getAmount()),
            true
        );

        System.out.print("Category (" + expense.getCategory() + "): ");
        String category = scanner.nextLine().trim();

        System.out.print("Description (" + expense.getDescription() + "): ");
        String description = scanner.nextLine().trim();

        String date = promptForDate("Date (" + expense.getDate() + "): ", true);

        if (amount != null) {
            expense.setAmount(amount);
        }
        if (!category.equals("")) {
            expense.setCategory(category);
        }
        if (!description.equals("")) {
            expense.setDescription(description);
        }
        if (!date.equals("")) {
            expense.setDate(date);
        }

        System.out.println("Expense updated.");
    }

    private void deleteExpense() {
        System.out.println("\nDelete Expense");
        int index = promptForExpenseIndex();

        if (index == -1) {
            return;
        }

        Expense expense = expenses.get(index);
        System.out.println("Delete " + expense + "?");
        System.out.print("Type yes to delete: ");
        String confirm = scanner.nextLine().trim().toLowerCase(Locale.ROOT);

        if (confirm.equals("yes")) {
            expenses.remove(index);
            System.out.println("Expense deleted.");
        } else {
            System.out.println("Delete canceled.");
        }
    }

    private void viewExpenses(List<Expense> expenseList) {
        viewExpenses(expenseList, "All Expenses");
    }

    private void viewExpenses(List<Expense> expenseList, String title) {
        System.out.println("\n" + title);

        if (expenseList.isEmpty()) {
            System.out.println("No expenses recorded.");
            return;
        }

        for (int i = 0; i < expenseList.size(); i++) {
            System.out.println((i + 1) + ". " + expenseList.get(i));
        }
    }

    private void searchByCategory() {
        System.out.println("\nSearch By Category");
        System.out.print("Enter category: ");
        String category = scanner.nextLine().trim().toLowerCase(Locale.ROOT);

        ArrayList<Expense> matches = new ArrayList<>();

        for (Expense expense : expenses) {
            if (expense.getCategory().toLowerCase(Locale.ROOT).equals(category)) {
                matches.add(expense);
            }
        }

        if (matches.isEmpty()) {
            System.out.println("No expenses found in that category.");
            return;
        }

        viewExpenses(matches, "Matching Expenses");
    }

    private void filterByDateRange() {
        System.out.println("\nFilter By Date Range");
        String startText = promptForDate("Start date (YYYY-MM-DD): ", false);
        String endText = promptForDate("End date (YYYY-MM-DD): ", false);
        LocalDate startDate = parseDate(startText);
        LocalDate endDate = parseDate(endText);

        if (startDate.isAfter(endDate)) {
            System.out.println("Start date cannot be after end date.");
            return;
        }

        ArrayList<Expense> matches = new ArrayList<>();

        for (Expense expense : expenses) {
            LocalDate expenseDate = parseDate(expense.getDate());

            if (expenseDate == null) {
                continue;
            }

            if (!expenseDate.isBefore(startDate) && !expenseDate.isAfter(endDate)) {
                matches.add(expense);
            }
        }

        if (matches.isEmpty()) {
            System.out.println("No expenses found in that date range.");
            return;
        }

        viewExpenses(matches, "Expenses From " + startText + " To " + endText);
    }

    private void sortExpenses() {
        System.out.println("\nSort Expenses");

        if (expenses.isEmpty()) {
            System.out.println("No expenses recorded.");
            return;
        }

        System.out.println("1. Date");
        System.out.println("2. Amount");
        System.out.println("3. Category");
        System.out.println("4. Description");
        System.out.print("Sort by: ");
        String choice = scanner.nextLine().trim();

        System.out.print("Descending? (y/n): ");
        boolean reverse = scanner.nextLine().trim().toLowerCase(Locale.ROOT).equals("y");

        Comparator<Expense> comparator;
        String title;

        if (choice.equals("1")) {
            comparator = Comparator.comparing(expense -> parseDate(expense.getDate()));
            title = "Expenses Sorted By Date";
        } else if (choice.equals("2")) {
            comparator = Comparator.comparingDouble(Expense::getAmount);
            title = "Expenses Sorted By Amount";
        } else if (choice.equals("3")) {
            comparator = Comparator.comparing(
                expense -> expense.getCategory().toLowerCase(Locale.ROOT)
            );
            title = "Expenses Sorted By Category";
        } else if (choice.equals("4")) {
            comparator = Comparator.comparing(
                expense -> expense.getDescription().toLowerCase(Locale.ROOT)
            );
            title = "Expenses Sorted By Description";
        } else {
            System.out.println("Invalid sort option.");



            return;
        }

        ArrayList<Expense> sortedExpenses = new ArrayList<>(expenses);

        if (reverse) {
            comparator = comparator.reversed();
        }

        sortedExpenses.sort(comparator);
        viewExpenses(sortedExpenses, title);
    }

    private void showTotal() {
        double total = 0;

        for (Expense expense : expenses) {
            total += expense.getAmount();
        }

        System.out.printf("%nTotal spent: $%.2f%n", total);
    }

    private void showTotalsByCategory() {
        System.out.println("\nTotals By Category");

        if (expenses.isEmpty()) {
            System.out.println("No expenses recorded. Please add expenses to see totals by category:)");
            return;
        }

        TreeMap<String, Double> totals = new TreeMap<>();

        for (Expense expense : expenses) {
            String category = expense.getCategory();
            double categoryTotal = totals.getOrDefault(category, 0.0);
            totals.put(category, categoryTotal + expense.getAmount());
        }

        for (Map.Entry<String, Double> entry : totals.entrySet()) {
            System.out.printf("%s: $%.2f%n", entry.getKey(), entry.getValue());
        }
    }

    private void saveExpenses() {
        File dataFile = new File(DATA_FILE);
        File parentFolder = dataFile.getParentFile();

        if (parentFolder != null) {
            parentFolder.mkdirs();
        }

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(dataFile))) {
            writer.write("amount,category,description,date");
            writer.newLine();

            for (Expense expense : expenses) {
                writer.write(expense.toCsvLine());
                writer.newLine();
            }

            System.out.println("Saved " + expenses.size() + " expenses to " + DATA_FILE + ".");
        } catch (IOException error) {
            System.out.println("Could not save expenses: " + error.getMessage());
        }
    }

    private void loadExpenses() {
        expenses.clear();

        File dataFile = new File(DATA_FILE);

        if (!dataFile.exists()) {
            return;
        }

        try (BufferedReader reader = new BufferedReader(new FileReader(dataFile))) {
            reader.readLine();
            String line;

            while ((line = reader.readLine()) != null) {
                String[] parts = splitCsvLine(line);

                if (parts.length != 4) {
                    continue;
                }

                try {
                    double amount = Double.parseDouble(parts[0]);
                    String category = parts[1].trim();
                    String description = parts[2].trim();
                    String date = parts[3].trim();

                    if (amount < 0 || category.equals("") || description.equals("")) {
                        continue;
                    }

                    if (parseDate(date) == null) {
                        continue;
                    }

                    expenses.add(new Expense(amount, category, description, date));
                } catch (NumberFormatException error) {
                    // Skip invalid rows so one bad line does not break the program.
                }
            }
        } catch (IOException error) {
            System.out.println("Could not load expenses: " + error.getMessage());
        }
    }

    private String[] splitCsvLine(String line) {
        ArrayList<String> values = new ArrayList<>();
        StringBuilder currentValue = new StringBuilder();
        boolean insideQuotes = false;

        for (int i = 0; i < line.length(); i++) {
            char currentChar = line.charAt(i);

            if (currentChar == '"') {
                if (insideQuotes && i + 1 < line.length() && line.charAt(i + 1) == '"') {
                    currentValue.append('"');
                    i++;
                } else {
                    insideQuotes = !insideQuotes;
                }
            } else if (currentChar == ',' && !insideQuotes) {
                values.add(currentValue.toString());
                currentValue.setLength(0);
            } else {
                currentValue.append(currentChar);
            }
        }

        values.add(currentValue.toString());
        return values.toArray(new String[0]);
    }
}
